lib/LogUtils.cc

```diff
@@ -25,6 +25,7 @@
 
 namespace pulsar {
 
+static std::atomic<LoggerFactory*> s_defaultLoggerFactory(new ConsoleLoggerFactory());
```

Comment:
The `s_defaultLoggerFactory` is never released. It's not an issue if the application only run once, but if this library is dynamically loaded by an external system, it will be a memory leak.

Could a static logger factory be used instead?

```c++
static ConsoleLoggerFactory s_defaultLoggerFactory;
```

Then return `&s_defaultLoggerFactory` in `getLoggerFactory()`.

Reply:
@BewareMyPower
`s_loggerFactory` also has the same issue, right? `s_loggerFactory` used to be a static shared_ptr, but it seems that it was changed to its current type to avoid segfaults during process shutdown.
https://github.com/apache/pulsar/pull/7132

Reply:
Should I fix it like this?
```cpp
static std::atomic<LoggerFactory*> s_defaultLoggerFactory(nullptr);

LoggerFactory* LogUtils::getLoggerFactory() {
    if (s_loggerFactory.load() != nullptr) {
        return s_loggerFactory.load();
    } else {
        if (s_defaultLoggerFactory.load() == nullptr) {
            LoggerFactory* oldFactory = nullptr;
            LoggerFactory* newFactory = new ConsoleLoggerFactory();
            if (!s_defaultLoggerFactory.compare_exchange_strong(oldFactory, newFactory)) {
                delete newFactory; // there's already a factory set
            }
        }
        return s_defaultLoggerFactory.load();
    }
}
```

Reply:
Oh I see. I think currently it's okay. Your fix here is also good.

Reply:
Then this patch LGTM no matter if this fix is applied.

Reply:
Thank you for your confirmation. Then, I will not make [the above modifications](https://github.com/apache/pulsar-client-cpp/pull/377#discussion_r1436381675).